src/nat/builder/child_builder.py

```
@@ -0,0 +1,338 @@
+        tools = await self._workflow_builder.get_tools(tool_names, wrapper_type)
+        for tool_name in tool_names:
+            if isinstance(self._workflow_builder, WorkflowBuilder):
+                function_groups = self._workflow_builder._function_groups
+            elif isinstance(self._workflow_builder, PerUserWorkflowBuilder):
+                # Per-user components can have dependencies on both shared and per-user function groups
+                function_groups = {
+                    **self._workflow_builder._shared_builder._function_groups,
+                    **self._workflow_builder._per_user_function_groups
+                }
+            else:
+                raise TypeError(f"Invalid workflow builder type: {type(self._workflow_builder)}")
```

Ruff (0.14.8)

Avoid specifying long messages outside the exception class

(TRY003)